connector/avro/src/main/scala/org/apache/spark/sql/avro/SchemaConverters.scala

```diff
@@ -142,18 +143,30 @@ object SchemaConverters {
         if (avroSchema.getTypes.asScala.exists(_.getType == NULL)) {
           // In case of a union with null, eliminate it and make a recursive call
           val remainingUnionTypes = AvroUtils.nonNullUnionBranches(avroSchema)
-          if (remainingUnionTypes.size == 1) {
-            toSqlTypeHelper(remainingUnionTypes.head, existingRecordNames, avroOptions)
-              .copy(nullable = true)
-          } else {
-            toSqlTypeHelper(
-              Schema.createUnion(remainingUnionTypes.asJava),
-              existingRecordNames,
-              avroOptions).copy(nullable = true)
-          }
+          toSqlTypeHelper(
+            Schema.createUnion(remainingUnionTypes.asJava),
+            existingRecordNames,
+            avroOptions).copy(nullable = true)
         } else avroSchema.getTypes.asScala.map(_.getType).toSeq match {
           case Seq(t1) =>
-            toSqlTypeHelper(avroSchema.getTypes.get(0), existingRecordNames, avroOptions)
+            // If spark.sql.avro.alwaysConvertUnionToStructType is set to false (default),
+            // we convert Avro union with a single primitive type into a primitive Spark type
+            // instead of a StructType.
+            if (!SQLConf.get.avroAlwaysConvertUnionToStruct) {
+              toSqlTypeHelper(avroSchema.getTypes.get(0), existingRecordNames, avroOptions)
```

Comment:
I was going to suggest doing this for Avro schema converter in the Avro data source where this is required because the top schema is always expected to be a struct.

For example, we change this code in AvroUtils: 
```scala
// Converts Avro schema to sql type and ensures that the top level data type is either an Avro
// record or a complex union that both result in a conversion to StructType
def convertAvroToSqlSchema(avroSchema: Schema, avroOptions: AvroOptions): StructType = {
  SchemaConverters.toSqlTypeHelper(avroSchema, Set.empty, avroOptions).dataType match {
    case t: StructType => t
    case _ => throw new RuntimeException(
      s"""Avro schema cannot be converted to a Spark SQL StructType:
         |
         |${avroSchema.toString(true)}
         |""".stripMargin)
  }
}
```

to something like this: 
```scala
// Converts Avro schema to sql type and ensures that the top level data type is either an Avro
// record or a complex union that both result in a conversion to StructType
def convertAvroToSqlSchema(avroSchema: Schema, avroOptions: AvroOptions): StructType = {
  SchemaConverters.toSqlTypeHelper(avroSchema, Set.empty, avroOptions).dataType match {
    case t: StructType => t
    case t: AtomicType => StructType(StructField("value", t, nullable = true))
    case _ => throw new RuntimeException(
      s"""Avro schema cannot be converted to a Spark SQL StructType:
         |
         |${avroSchema.toString(true)}
         |""".stripMargin)
  }
}
```

Can we also check that this change will not affect `from_avro` SQL function?

Reply:
Maybe it is good to compare the two approaches and see which one makes sense.

Reply:
I don't think the two approaches are equivalent.

```scala
SchemaConverters.toSqlType(avroSchema, options).dataType match {
  case t: StructType => Some(t)
  case _ => throw new RuntimeException(
    s"""Avro schema cannot be converted to a Spark SQL StructType:
       |
       |${avroSchema.toString(true)}
       |""".stripMargin)
}
``` 
is for converting Avro top-level records into Spark StructType IIUC. In deserialization, we use `getRecordWriter` to wrap around Avro schema that are non-records into records, and in this method we get it back.

Reply:
When we call this method, we expect the datatype to be `StructType` already, and that is why all other matches would throw exceptions.

Reply:
My point is that adding a case match for `AtomicType` here:
1.  would not help us convert Avro union of a single type to StructType
2. would emit exceptions that should be thrown here and have undesired behaviors

Reply:
Sorry, I did not get what the problem was with my suggested approach. Could you clarify what would fail next if you had made the above change?

Reply:
The major problem is that the check is meant to safeguard that the top level Avro schema is either an Avro
record or a complex union. 
If we add `case t: AtomicType => StructType(StructField("value", t, nullable = true))`, it loosen the restriction and the top level schema could just be an integer or something.